### PR TITLE
Use rust gates for Optimize1QGatesDecomposition

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit moves to using rust gates for the Optimize1QGatesDecomposition transpiler pass. It takes in a sequence of runs (which are a list of DAGOpNodes) from the python side of the transpiler pass which are generated from DAGCircuit.collect_1q_runs() (which in the future should be moved to rust after #12550 merges). The rust portion of the pass now iterates over each run, performs the matrix multiplication to compute the unitary of the run, then synthesizes that unitary, computes the estimated error of the circuit synthesis and returns a tuple of the circuit sequence in terms of rust StandardGate enums. The python portion of the code then takes those sequences and does inplace substitution of each run with the sequence returned from rust.

Once #12550 merges we should be able to move the input collect_1q_runs() call and perform the output node substitions in rust making the full pass execute in the rust domain without any python interaction.

Additionally, the OneQubitEulerDecomposer class is updated to use rust for circuit generation instead of doing this python side. The internal changes done to use rust gates in the transpiler pass meant we were half way to this already by emitting rust StandardGates instead of python gate objects. The dag handling is still done in Python however until #12550 merges.

~This also includes an implementation of the r gate, I temporarily added this to unblock this effort as it was the only gate missing needed to complete this. We can rebase this if a standalone implementation of the gate merges before this.~ Rebased on top of: https://github.com/Qiskit/qiskit/pull/12662

### Details and comments

TODO:

- [x] Investigate multithreading across runs. This might not make sense until #12550 merges so we can do a parallel bridge iterator over runs and avoid having to collect runs generated in python

### CODE REVIEW

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: I need to add back the caching mechanism here so we don't need to recompute this for every qubit.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]:
```suggestion
```

**Comment by mtreinish** [comment]
I took a quick attempt at implementing this with multithreading locally and I could get it to work but it required doing extra collection of all the data from the collect runs into a second vec to strip the `PyRef` from each node. This added more overhead than parallelism saved in this case. I think it should be possible to investigate doing this in parallel after #12550 merges because then we can generate the runs from rust directly. For multithreading we'll likely still need to collect the runs into a vec (to avoid mutation while iterating) but we can measure if that overhead is higher than the speedup from doing the synthesis in parallel and investigate in a followup.

**Comment by mtreinish** [comment]
I threw together a quick benchmark to test this:

```python
import statistics
import time

from qiskit.transpiler.passes import Optimize1qGatesDecomposition
from qiskit.circuit.random import random_circuit
from qiskit.converters import circuit_to_dag

seed = 42
n_qubits = 20
depth = 102400

qc = random_circuit(n_qubits, depth, measure=True, seed=seed)
dag = circuit_to_dag(qc)
basis_gates = ['sx', 'x', 'rz', 'cx', 'id']
opt_pass = Optimize1qGatesDecomposition(basis=basis_gates)
times = []
for i in range(10):
    start = time.perf_counter()
    opt_pass.run(dag)
    stop = time.perf_counter()
    times.append(stop - start)
print(f"Mean runtime: {statistics.geometric_mean(times)}")
```

and it's showing a pretty noticeable speedup over Qiskit 1.1. With this PR I got:

```
3.931521232593536 sec
```

and with Qiskit 1.1 I got:

```
6.184648169533495 sec
```

Looking at a profile with this script the majority of the time with this PR is spent in collect 1q runs and dag interactions in python.

**Comment by mtreinish** [comment]
To compare this against the current state of main I ran asv for the pass specific benchmarks:

```
Benchmarks that have improved:

| Change   | Before [1ed5951a] <optimize-1q-more-rust~3^2>   | After [130969a3] <optimize-1q-more-rust>   |   Ratio | Benchmark (Parameter)                                                                                         |
|----------|-------------------------------------------------|--------------------------------------------|---------|---------------------------------------------------------------------------------------------------------------|
| -        | 134±1ms                                         | 46.9±0.4ms                                 |    0.35 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
| -        | 99.3±1ms                                        | 34.0±0.2ms                                 |    0.34 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
| -        | 116±0.7ms                                       | 39.0±0.5ms                                 |    0.34 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(20, 1024, ['u', 'cx', 'id'])                    |
| -        | 85.6±0.3ms                                      | 27.9±0.1ms                                 |    0.33 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(14, 1024, ['u', 'cx', 'id'])                    |
| -        | 123±1ms                                         | 41.1±0.5ms                                 |    0.33 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(20, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
| -        | 92.0±0.5ms                                      | 29.6±0.2ms                                 |    0.32 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
| -        | 47.4±0.5ms                                      | 12.6±0.04ms                                |    0.27 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
| -        | 43.6±0.3ms                                      | 10.6±0.05ms                                |    0.24 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
| -        | 40.7±0.4ms                                      | 9.31±0.06ms                                |    0.23 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_decompose(5, 1024, ['u', 'cx', 'id'])                     |

Benchmarks that have stayed the same:

| Change   | Before [1ed5951a] <optimize-1q-more-rust~3^2>   | After [130969a3] <optimize-1q-more-rust>   |   Ratio | Benchmark (Parameter)                                                                                           |
|----------|-------------------------------------------------|--------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------|
|          | 2.65±0.01s                                      | 2.66±0.02s                                 |    1    | passes.PassBenchmarks.time_optimize_swap_before_measure(20, 1024)                                               |
|          | 231±1ms                                         | 230±1ms                                    |    1    | passes.PassBenchmarks.time_optimize_swap_before_measure(5, 1024)                                                |
|          | 737±10ms                                        | 719±6ms                                    |    0.98 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 1.53±0.01s                                      | 1.51±0s                                    |    0.98 | passes.PassBenchmarks.time_optimize_swap_before_measure(14, 1024)                                               |
|          | 1.14±0.01s                                      | 1.11±0s                                    |    0.97 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
|          | 256±2ms                                         | 247±1ms                                    |    0.97 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
|          | 822±10ms                                        | 778±3ms                                    |    0.95 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 746±7ms                                         | 706±2ms                                    |    0.95 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(14, 1024, ['u', 'cx', 'id'])                    |
|          | 1.26±0.01s                                      | 1.20±0s                                    |    0.95 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
|          | 1.16±0.02s                                      | 1.10±0s                                    |    0.95 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(20, 1024, ['u', 'cx', 'id'])                    |
|          | 298±2ms                                         | 277±1ms                                    |    0.93 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
|          | 260±2ms                                         | 239±1ms                                    |    0.92 | passes.MultipleBasisPassBenchmarks.time_optimize_1q_commutation(5, 1024, ['u', 'cx', 'id'])                     |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (ElePT) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs:701]: why is `gates.len()` returned twice here? I would expect this output to be something like `None => (None, gates.len()),`

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (ElePT) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]: Is there any difference between `basis.deref()` and `*basis`?

**Review by ElePT — COMMENTED** [review]
I tried taking a look at the implementation, I didn't see anything that stood out as wrong but I'm sure I wasn't able to catch every detail.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:701]: This is for compatibility with the python side scoring. The scoring function in python are designed to do sorting with a tuple of `(error_rate, len)` and in the case without error rates we just use `(len, len)`. I just retained this from what the python side was doing before.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: Using `*basis` would end up in a copy I think but this does it all by reference. Although I'm not sure it would work because basis is a `PyBackedStr` so I don't think we can use `*basis`. `PyBackedStr.deref()`  returns a `&str` which is what we need in this case.

**Comment by mtreinish** [comment]
We should merge this after #12662 merges since this depends on having an RGate implementation in rust. I did it inline here but it would be good to review that in isolation, I'll rebase this after that merges.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]: Very minor, but would save looking upward a few lines for context. I tried with rustfmt; no change in linewrap!
```suggestion
                .map(|gate| 1. - err_map.get(gate.0.name()).unwrap_or(&0.))
```

Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: Have we considered using a non-string key? (Or some kind of [optimized string](https://swatinem.de/blog/optimized-strings/))

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]:
```suggestion
                .map(|gate| 1. - err_map.error_map[qubit].get(gate.0.name()).unwrap_or(&0.))
```
Also no fmt change.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]:
```suggestion
                .map(|gate| compute_error_term(gate.0.as_str(), err_map, qubit))
```
No change with `rustfmt`

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: Not really, these strings are lookup keys from the `Target`, the error map here is just a rust view of the error data in the target that we construct in Python to provide the data to this pass (when we have #12292 merged we can drop this structure and query it directly). We need something that interoperates with a Python string which is realistically just `String` or `&str`.

Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: I assumed that, at least for now, we need regular strings for Python interop.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]: EDIT: This should have the same behavior.
```suggestion
    let target_basis_vec: PyResult<Vec<EulerBasis>> = target_basis_list
        .iter()
        .map(|basis| EulerBasis::__new__(basis))
        .collect();
    Ok(unitary_to_gate_sequence_inner(
        unitary.as_array(),
        &target_basis_vec?,
```

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]:
```suggestion
            let target_basis_vec: Vec<EulerBasis> = bases[index]
                .iter()
                .map(|basis| EulerBasis::__new__(basis).unwrap())
                .collect();
```

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: [Here is a diff](https://github.com/mtreinish/qiskit-core/compare/optimize-1q-more-rust...jlapeyre:qiskit-core:optimize-1q-more-rust-gjl-change) showing  everywhere the pattern occurs

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: I might be mistaken, but the docs suggest that `PyBackedStr` is meant to be used with deref coercion. You can use `__new__(basis)`.  [For example.](https://github.com/mtreinish/qiskit-core/compare/optimize-1q-more-rust...jlapeyre:qiskit-core:optimize-1q-more-rust-gjl-change)

**Comment by jlapeyre** [comment]
In the code in https://github.com/Qiskit/qiskit/pull/12650#issuecomment-2189462949 inserting nonsense for `basis_gates`, eg `basis_gates = ['NONSENSE', 'x', 'rz', 'cx', 'id']` throws no error. The code runs to completion silently (I made the sizes and reps small so it runs faster)

Of course, the same thing happens on the `main` branch. Maybe there is a good reason.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: I applied this to the branch: https://github.com/Qiskit/qiskit/pull/12650/commits/68af440ca1ad08d3a0551c7e00013eddb29b2434

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: I applied this suggestion to the branch in: https://github.com/Qiskit/qiskit/pull/12650/commits/68af440ca1ad08d3a0551c7e00013eddb29b2434

**Comment by mtreinish** [comment]
> In the code in https://github.com/Qiskit/qiskit/pull/12650#issuecomment-2189462949 inserting nonsense for basis_gates, eg basis_gates = ['NONSENSE', 'x', 'rz', 'cx', 'id'] throws no error. The code runs to completion silently (I made the sizes and reps small so it runs faster)
>
> Of course, the same thing happens on the main branch. Maybe there is a good reason.

That's expected behavior. The target euler basis is selected as a subset of the provided basis gate names. So it looks for which gate names it knows about and figures out which are valid decomposition bases to use. So in this context `cx` is treated the same way as `NONSENSE`. If we were using a target we'd have to associate `NONSENSE` with an actual python gate definition. But  this pass would still work the same way because the decomposition is only defined for a fixed set of target gates. You can see the table here: https://docs.quantum.ibm.com/api/qiskit/qiskit.synthesis.OneQubitEulerDecomposer

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: EDIT: No not faster.
I tried this, thinking it might be faster:
```rust
fn matmul_1q(operator: &mut [[Complex64; 2]; 2], other: Array2<Complex64>) {
    let (a00, a01, a10, a11) = (operator[0][0], operator[0][1], operator[1][0], operator[1][1]);
    operator[0][0] =  other[[0, 0]] * a00 + other[[0, 1]] * a10;
    operator[0][1] =  other[[0, 0]] * a01 + other[[0, 1]] * a11;
    operator[1][0] =  other[[1, 0]] * a00 + other[[1, 1]] * a10;
    operator[1][1] =  other[[1, 0]] * a01 + other[[1, 1]] * a11;
}
```
Perhaps this avoids an allocation.
But performance of this code https://github.com/Qiskit/qiskit/pull/12650#issuecomment-2189462949 did not appear to change.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]: There may be a better message, but something like:
```suggestion
                        .expect("No matrix defined for operation")
```
My understanding is `expect` will only panic here if there is a programming error, not a user error; and that this is a legitimate use of `expect`. Not great error handling, but a step up from `unwrap`.

If we follow the rust doc style guidelines, the message should be something like `"Operation must have a matrix definition"`. IDK which is preferrable.

We could also do `unwrap_or_else(|| panic!("{}", format!("No matrix defined for operation {}", node.instruction.operation.name()` (untested)

Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: Sure, I can add an `expect()` call here. It shouldn't be possible to reach this code and not have a matrix defined since the input nodes to this function have passed through this filter function: https://github.com/Qiskit/qiskit/pull/12650/files#diff-61c072175672ef132e5e682a51401abd08288938d864b2090f4dc446e772837eR1048-R1069 (which checks that `matrix()` `is_some()`)

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: I'm not sure where this would save an allocation this is just destructuring the array reference to individual variables. I would expect the compiler to generate the same exact code for this as the code in this function already.

The one allocation point I'm not happy about but I couldn't really figure out how to avoid was around the `Array2` for `other` which we're getting from `Operation::matrix()`. I needed to do that from a typing perspective in the `Operation` trait, because using an array view wouldn't work since we may get the array view from Python or it might point to a static array so I couldn't see a way to declare a lifetime parameter for the trait method that would work (or at least I think that's why it returns an owned array, it's been >1 month since I wrote that class so I'm not 100%).

Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: Another thing. We can test if all gates are standard, and if so do some optimizations. Like use fixed sized `Array`s throughout the multiplication. Doesn't need to be in this PR of course.

Comment (mtreinish) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: Yeah, let's look at that kind of thing in a followup. Right now the runtime overhead is mostly dominated by the rust python boundary and I think we'll have another round or two of optimizations to perform on this pass after #12550 merges. We also should have more flexibility when everything is handled in rust.

Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: > just destructuring the array reference to individual variables. 

I see that now. In fact, if there were an allocation, there is no way the caller can know where it is.

Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs:1034]: > avoid was around the `Array2` for other which we're getting from `Operation::matrix()`

This is the optimization I was talking about above. I didn't think it through all the way. But if you could scan or use a flag to know that all matrices are slices of fixed size, then multiplication would be much faster. They are all 2x2 of course, but `Array2` doesnt carry the info in the type.
But i wouldnt want to slow this PR down for this optimization.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [reply crates/accelerate/src/euler_one_qubit_decomposer.rs]: Right. It's meant to signal somehow that this is unreachable unless something is broken, ie a programming error creeps in. People debate over `expect`, `unreachable!`, etc. I like something better than `unwrap` anyway.  You could ask, is the programmer not handling an error that might arise with bad input? In part this is meant to signal to the reader that this will only fail (we believe) if someone breaks code elsewhere.

We could even write "you have discovered a bug in qiskit", etc. But I think something minimally visually distracting is good.

**Comment by jlapeyre** [comment]
Another note for later. This function
https://github.com/Qiskit/qiskit/blob/3fb175390d5fa007150fb76a23b914059010788a/crates/accelerate/src/euler_one_qubit_decomposer.rs#L386-L386
has return type `PyResult<OneQubitGateSequence>`. But as far as I can tell it is only called in one place, and that is in the same file. The returned value is `unwrap`ped with no check. In fact the return value is always `ok`. This is so the function can be called from Python. (maybe in the future, since I don't see this called anywhere) Reading and understanding the `unwrap` takes a bit of time.

I'd prefer that the Python-side function (if needed) wraps a function that returns `OneQubitGateSequence` rather than having this special requirement of the Python interface intrude in code that is only used in this function.

This is a quick fix. But probably outside the scope of this PR.

File: crates/accelerate/src/euler_one_qubit_decomposer.rs
Comment (jlapeyre) [inline crates/accelerate/src/euler_one_qubit_decomposer.rs]:
```suggestion
                && match &node.instruction.extra_attrs {
                    None => true,
                    Some(attrs) => attrs.condition.is_none(),
                }
```

**Comment by jlapeyre** [comment]
I left one more suggested simplification edit.

Other than that, this looks good to go.

**Review by jlapeyre — APPROVED** [review]

**Pull request merged**
